File: 03_sorting/03_insertion_sort/solution.py
# insertion sort
class Solution:
    def insertionSort(self, arr):
        sorted_till = 0
        
        while sorted_till < len(arr) - 1:
            # the next index whose value has to be inserted in sorted array
            next = sorted_till + 1
            
            while next > 0 and arr[next] < arr[next-1]:
                arr[next], arr[next-1] = arr[next-1], arr[next]
                next-=1
            
            # we sorted one more element in the array 
            sorted_till += 1

File: 03_sorting/03_insertion_sort/solution.java
// insertion sort
class Solution {
    public void insertionSort(int arr[]) {
        int sorted_till = 0;
        
        while (sorted_till < arr.length - 1) {
            // take next-index whose value needs to be inserted
            int next = sorted_till + 1;
            
            while ((next > 0) && (arr[next] < arr[next-1])) {
                // swap values at "next" and "next-1"
                int temp = arr[next];
                arr[next] = arr[next-1];
                arr[next-1] = temp;
                
                next -= 1;
            }
            
            // at end of while loop the sorted subset array length increases by 1
            sorted_till += 1;
        }
    }
}